manager/probe.go

```diff
@@ -532,46 +533,68 @@ func (p *Probe) reset() {
 	p.checkPin = false
 	p.funcName = ""
 	p.attachPID = 0
+	p.attachedWithSysfs = false
 }
 
 // attachKprobe - Attaches the probe to its kprobe
 func (p *Probe) attachKprobe() error {
 	// Prepare kprobe_events line parameters
-	var probeType, maxactiveStr string
+	var probeType, maxactiveStr, sectionPrefix string
 	var err error
 	funcName := p.funcName
-	if strings.HasPrefix(p.Section, "kretprobe/") {
+
+	dividedSection := strings.Split(p.Section, "/")
+	if len(dividedSection) >= 1 {
+		sectionPrefix = dividedSection[0]
+	} else {
+		return errors.Wrapf(ErrSectionFormat, "program type unrecognized in section %v", p.Section)
+	}
+
+	switch sectionPrefix {
+	case "kretprobe":
 		if funcName == "" {
 			funcName = strings.TrimPrefix(p.Section, "kretprobe/")
 		}
 		if p.KProbeMaxActive > 0 {
 			maxactiveStr = fmt.Sprintf("%d", p.KProbeMaxActive)
 		}
 		probeType = "r"
-	} else if strings.HasPrefix(p.Section, "kprobe/") {
+	case "kprobe":
 		if funcName == "" {
 			funcName = strings.TrimPrefix(p.Section, "kprobe/")
 		}
 		probeType = "p"
-	} else {
+	default:
 		// this might actually be a Uprobe
 		return p.attachUprobe()
 	}
 	p.attachPID = os.Getpid()
 
-	// Write kprobe_events line to register kprobe
-	kprobeID, err := EnableKprobeEvent(probeType, funcName, p.UID, maxactiveStr, p.attachPID)
-	// fallback without KProbeMaxActive
-	if err == ErrKprobeIDNotExist {
-		kprobeID, err = EnableKprobeEvent(probeType, funcName, p.UID, "", p.attachPID)
-	}
+	// Try to use the perf_event_open API first (e12f03d "perf/core: Implement the 'perf_kprobe' PMU")
+	perfEventFD, err := perfEventOpenWithProbe(funcName, 0, -1, sectionPrefix, 0)
```

Comment:
Do we know how this reacts when the kernel doesn't support it? Rather than attempt this for every probe attach, it would be great to disable it when not available, such as kernels < 4.17

Reply:
We could maybe try using an `internal.FeatureTest` for this.

Reply:
Yeah it simply fails and returns an error ... I'm simply following BCC design again, but I'll try to figure out a cleaner way of dealing with this and get back to you 😄